File: examples/ts/http/agent.ts
/**
 * TypeScript HTTP transport Lilith example — Streamable HTTP MCP 2025-11-25.
 *
 * Mirrors examples/python/http/agent.py exactly:
 *   1. Start the local MCP server (server.mjs) in-process
 *   2. Connect Lilith with upstreamUrl → --transport http
 *   3. Tool discovery, allowed call, denied call → PolicyViolationError
 *   4. Session ID + full audit log
 *
 * Run:
 *   export LILITH_ZERO_BINARY_PATH=/path/to/lilith-zero
 *   bun agent.ts
 */

import { join } from "path";
import { Lilith, PolicyViolationError } from "../../../sdk-ts/src/index.ts";
// @ts-ignore — .mjs module, no types needed
import { serve, HOST, PORT } from "./server.mjs";

const POLICY = join(import.meta.dir, "policy.yaml");
const SERVER_URL = `http://${HOST}:${PORT}/mcp`;

// Start the HTTP MCP server in-process.
const httpServer = serve();
// Give the socket a moment to bind.
await new Promise<void>((r) => setTimeout(r, 50));

const lilith = new Lilith({ upstreamUrl: SERVER_URL, policy: POLICY });
await using _ = lilith;
await lilith.connect();

console.log(`transport : HTTP → ${SERVER_URL}`);
console.log(`session   : ${lilith.sessionId}`);

const tools = await lilith.listTools();
console.log(`tools     : [${tools.map((t) => t.name).join(", ")}]`);

const search = await lilith.callTool("search_web", { query: "MCP security" });
console.log(`search    : ${(search.content[0] as { text: string }).text}`);

const time = await lilith.callTool("get_time", {});
console.log(`time      : ${(time.content[0] as { text: string }).text}`);

try {
  await lilith.callTool("query_database", { sql: "SELECT * FROM users" });
  console.log("ERROR: should have been blocked");
} catch (e) {
  if (e instanceof PolicyViolationError) {
    console.log(`blocked   : ${e.message}`);
  } else throw e;
}

const audit = await lilith.drainAuditLogs();
console.log(`\naudit     : ${audit.length} entries, ${audit.filter((e) => e.signature).length} signed`);
for (const entry of audit) {
  const dec = (entry.details["decision"] as string | undefined) ?? entry.event_type;
  const tool = (entry.details["tool_name"] as string | undefined) ?? "";
  console.log(`  [${dec.padEnd(12)}] ${tool}`);
}

httpServer.close();

File: examples/python/http/agent.py
"""
HTTP transport Lilith reference implementation.

Demonstrates Lilith proxying a Streamable HTTP MCP server (2025-11-25) instead
of a stdio child process.  Policy enforcement is identical to the stdio case —
the transport is the only difference.

What this shows:
  1. Start a local HTTP MCP server (server.py)
  2. Connect Lilith with --transport http → --upstream-url
  3. list_tools, allowed call, denied call (PolicyViolationError)
  4. Session ID and audit log — same guarantees as stdio

Run:
    export LILITH_ZERO_BINARY_PATH=/path/to/lilith-zero
    python agent.py
"""

import asyncio
import os
import sys
import time
from threading import Thread

from lilith_zero import Lilith, PolicyViolationError

# Import the embedded server so we can start it in-process.
sys.path.insert(0, os.path.dirname(__file__))
import server as _srv

POLICY = os.path.join(os.path.dirname(__file__), "policy.yaml")
SERVER_URL = f"http://{_srv.HOST}:{_srv.PORT}/mcp"


async def main() -> None:
    # Start the HTTP MCP server in a background daemon thread.
    httpd = _srv.serve(daemon=True)
    time.sleep(0.1)  # let the socket bind

    try:
        async with Lilith(
            upstream_url=SERVER_URL,
            policy=POLICY,
        ) as lilith:
            print(f"transport : HTTP → {SERVER_URL}")
            print(f"session   : {lilith.session_id}")

            tools = await lilith.list_tools()
            print(f"tools     : {[t['name'] for t in tools]}")

            result = await lilith.call_tool("search_web", {"query": "MCP security"})
            print(f"search    : {result['content'][0]['text']}")

            result = await lilith.call_tool("get_time", {})
            print(f"time      : {result['content'][0]['text']}")

            try:
                await lilith.call_tool("query_database", {"sql": "SELECT * FROM users"})
                print("ERROR: should have been blocked")
            except PolicyViolationError as exc:
                print(f"blocked   : {exc}")

            audit = await lilith.drain_audit_logs()
            print(f"\naudit     : {len(audit)} entries, "
                  f"{sum(1 for e in audit if e.get('signature'))} signed")
            for e in audit:
                dec = e.get("details", {}).get("decision", e.get("event_type", "?"))
                tool = e.get("details", {}).get("tool_name", "")
                print(f"  [{dec:<6}] {tool}")

    finally:
        httpd.shutdown()


if __name__ == "__main__":
    asyncio.run(main())
